# Data Preprocessing

In [ ]:
from keras.preprocessing import sequence
from keras import models, layers, optimizers, datasets, utils, losses

vocabulary_size = 10000
maxlen = 40
batch_size = 25

(x_train, y_train), (x_test, y_test) = datasets.imdb.load_data(num_words=vocabulary_size)
x_train = sequence.pad_sequences(x_train, maxlen)
x_test = sequence.pad_sequences(x_test, maxlen)

#LSTM

In [ ]:
inputs = layers.Input(shape=(maxlen,))
e=layers.Embedding(vocabulary_size, 128)(inputs)
h=layers.LSTM(128, dropout=0.8, recurrent_dropout=0.8)(e)
outputs=layers.Dense(1, activation='sigmoid')(h)
model_LSTM = models.Model(inputs, outputs)

model_LSTM.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

model_LSTM.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=4,
          validation_data=(x_test, y_test))

score_LSTM, acc_LSTM = model_LSTM.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test accuracy:', acc_LSTM)

Epoch 1/4
1000/1000 [==============================] - 93s 93ms/step - loss: 0.5240 - accuracy: 0.7276 - val_loss: 0.4212 - val_accuracy: 0.8054
Epoch 2/4
1000/1000 [==============================] - 94s 94ms/step - loss: 0.4032 - accuracy: 0.8164 - val_loss: 0.4236 - val_accuracy: 0.8058
Epoch 3/4
1000/1000 [==============================] - 91s 91ms/step - loss: 0.3608 - accuracy: 0.8405 - val_loss: 0.4366 - val_accuracy: 0.8064
Epoch 4/4
1000/1000 [==============================] - 14s 14ms/step - loss: 0.4532 - accuracy: 0.8056
Test accuracy: 0.8055599927902222


#GRU

In [ ]:
inputs = layers.Input(shape=(maxlen,))
e=layers.Embedding(vocabulary_size, 128)(inputs)
h=layers.GRU(128, dropout=0.8, recurrent_dropout=0.8)(e)
#h=layers.LSTM(128, dropout=0.8, recurrent_dropout=0.8)(e)
outputs=layers.Dense(1, activation='sigmoid')(h)
model_GRU = models.Model(inputs, outputs)

model_GRU.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

model_GRU.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=4,
          validation_data=(x_test, y_test))

score_GRU, acc_GRU = model_GRU.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test accuracy:', acc_GRU)

Epoch 1/4
1000/1000 [==============================] - 85s 85ms/step - loss: 0.5486 - accuracy: 0.7082 - val_loss: 0.4275 - val_accuracy: 0.8027
Epoch 2/4
1000/1000 [==============================] - 86s 86ms/step - loss: 0.4124 - accuracy: 0.8141 - val_loss: 0.4094 - val_accuracy: 0.8112
Epoch 3/4
1000/1000 [==============================] - 86s 86ms/step - loss: 0.3625 - accuracy: 0.8382 - val_loss: 0.4028 - val_accuracy: 0.8158
Epoch 4/4
1000/1000 [==============================] - 14s 14ms/step - loss: 0.4116 - accuracy: 0.8146
Test accuracy: 0.8146399855613708


#Dense

In [ ]:
inputs = layers.Input(shape=(maxlen,))
e=layers.Embedding(vocabulary_size, 128)(inputs)
h=layers.Dense(128)(e)
# h=layers.GRU(128, dropout=0.8, recurrent_dropout=0.8)(e)
# h=layers.LSTM(128, dropout=0.8, recurrent_dropout=0.8)(e)
outputs=layers.Dense(1, activation='sigmoid')(h)
model_Dense1 = models.Model(inputs, outputs)

model_Dense1.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

model_Dense1.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=4,
          validation_data=(x_test, y_test))

score_Dense1, acc_Dense1 = model_Dense1.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test accuracy:', acc_Dense1)

Epoch 1/4
1000/1000 [==============================] - 21s 21ms/step - loss: 0.6800 - accuracy: 0.5503 - val_loss: 0.6774 - val_accuracy: 0.5536
Epoch 2/4
1000/1000 [==============================] - 20s 20ms/step - loss: 0.6736 - accuracy: 0.5609 - val_loss: 0.6774 - val_accuracy: 0.5541
Epoch 3/4
1000/1000 [==============================] - 20s 20ms/step - loss: 0.6723 - accuracy: 0.5625 - val_loss: 0.6778 - val_accuracy: 0.5570
Epoch 4/4
1000/1000 [==============================] - 2s 2ms/step - loss: 0.6780 - accuracy: 0.5568
Test accuracy: 0.556821882724762


#Dense with dropout

In [ ]:
inputs = layers.Input(shape=(maxlen,))
e=layers.Embedding(vocabulary_size, 128)(inputs)
h=layers.Dense(128)(e)
d=layers.Dropout(0.8)(h)
# h=layers.GRU(128, dropout=0.8, recurrent_dropout=0.8)(e)
# h=layers.LSTM(128, dropout=0.8, recurrent_dropout=0.8)(e)
outputs=layers.Dense(1, activation='sigmoid')(d)
model_Dense2 = models.Model(inputs, outputs)

model_Dense2.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

model_Dense2.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=4,
          validation_data=(x_test, y_test))

score_Dense2, acc_Dense2 = model_Dense2.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test accuracy:', acc_Dense2)

Epoch 1/4
1000/1000 [==============================] - 20s 20ms/step - loss: 0.6813 - accuracy: 0.5487 - val_loss: 0.6776 - val_accuracy: 0.5563
Epoch 2/4
1000/1000 [==============================] - 20s 20ms/step - loss: 0.6748 - accuracy: 0.5594 - val_loss: 0.6775 - val_accuracy: 0.5569
Epoch 3/4
1000/1000 [==============================] - 20s 20ms/step - loss: 0.6730 - accuracy: 0.5623 - val_loss: 0.6777 - val_accuracy: 0.5569
Epoch 4/4
1000/1000 [==============================] - 2s 2ms/step - loss: 0.6781 - accuracy: 0.5539
Test accuracy: 0.5538750290870667


#Naive Bayes

In [ ]:
from sklearn.naive_bayes import MultinomialNB
from sklearn import metrics

model_Naive = MultinomialNB()
model_Naive.fit(x_train, y_train)
acc_Naive = model_Naive.score(x_test,y_test)
print('Test accuracy:', acc_Naive)

Test accuracy: 0.50704


#Picking three random word embedding vector

In [ ]:
import random
word1=embeddings[random.randint(0, embeddings.shape[0])]
word2=embeddings[random.randint(0, embeddings.shape[0])]
word3=embeddings[random.randint(0, embeddings.shape[0])]

#Euclidean Distances

In [ ]:
import numpy as np
print(np.linalg.norm(word1-word2))
print(np.linalg.norm(word1-word3))
print(np.linalg.norm(word2-word3))

0.44555712
0.45807117
0.4590523


* The Text Accuracy for LSTM is 80% whereas GRU is 81%. Not a significant change but they both are different versions so the accuracies differ slightly.
* We took 10K words, this may be enough for the LSTM and GRU to push accuracies to 80% because usually they are used if we have less data to train. 
* The major reason behind this accuracy is we are mitigating the vanishing gradient problem, which leads to better training of model and gives better predictions. 
* The accuracies of Dense layer is around 55%, very low because it will be normal RNN which has vanishing gradient problem and the first layers don't train well. This Dense version is not able to memorize the words that are entered at first in time. So, the model has to decide only on last time cycles which makes it difficult. 
* The Naive Bayes accuracy is also 50% which is very low. Naive Bayes will use conditional probability so it is not able to find a perfect condition for each word in sentense whenever detected.

